File: app/src/main/java/com/iomt/android/BLESearcher.java
package com.iomt.android;

import android.Manifest;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.bluetooth.le.BluetoothLeScanner;
import android.bluetooth.le.ScanCallback;
import android.bluetooth.le.ScanResult;
import android.content.pm.PackageManager;
import android.os.Build;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class BLESearcher extends AppCompatActivity implements SavedDeviceAdapter.OnClickListener {
    private Handler uiHandler;

    private MenuItem menuItem;

    private List<DeviceCell> deviceCells = new ArrayList<>();
    private static final int PERMISSION_REQUEST_FINE_LOCATION = 1;
    private SavedDeviceAdapter myAdapter;
    private BluetoothLeScanner mBluetoothLeScanner;
    private List<DeviceType> dev_types = new ArrayList<>();
    private List<DeviceInfo> devs = new ArrayList<>();
    private HTTPRequests httpRequests;


    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_searcher);


        String JWT = getIntent().getStringExtra("JWT");
        String userId = getIntent().getStringExtra("UserId");
        httpRequests = new HTTPRequests(this, JWT, userId);
        Action action_types = (String[] args) -> {
            GsonBuilder builder = new GsonBuilder();
            Gson gson = builder.create();
            Type listOfDeviceTypes = new TypeToken<ArrayList<DeviceType>>(){}.getType();
            dev_types = gson.fromJson(args[0], listOfDeviceTypes);
        };
        httpRequests.getDeviceTypes(action_types);

        Action action_devs = (String[] args) -> {
            GsonBuilder builder = new GsonBuilder();
            Gson gson = builder.create();
            Type listOfDevices= new TypeToken<ArrayList<DeviceInfo>>(){}.getType();
            devs = gson.fromJson(args[0], listOfDevices);
        };
        httpRequests.getDevices(action_devs);

        BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        mBluetoothLeScanner = bluetoothAdapter.getBluetoothLeScanner();
        uiHandler = new Handler(Looper.getMainLooper());


        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar_searcher);
        setSupportActionBar(toolbar);


        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.searcher_recycler_view);
        recyclerView.setHasFixedSize(true);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(layoutManager);
        myAdapter = new SavedDeviceAdapter(LayoutInflater.from(this), deviceCells, this);
        recyclerView.setAdapter(myAdapter);

        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(new String[] { Manifest.permission.ACCESS_FINE_LOCATION }, PERMISSION_REQUEST_FINE_LOCATION);
        }
    }

    private String get_type(BluetoothDevice d) {
        for (DeviceType i: dev_types) {
            if (d.getName().startsWith(i.getPrefix())) {
                return i.getDevice_type();
            }
        }
        return null;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        if (requestCode == PERMISSION_REQUEST_FINE_LOCATION) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                menuItem.setEnabled(true);
            } else {
                final AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("Location not granted");
                builder.setMessage("Since location access has not been granted, this app will not be able to discover nearby bluetooth devices.");
                builder.setPositiveButton(android.R.string.ok, null);
                builder.show();

                menuItem.setEnabled(false);
            }
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.menu_searcher, menu);
        menuItem = menu.findItem(R.id.scan);
        return true;
    }

    @Override
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.scan) {
            scanLeDevice();
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private ScanCallback mLeScanCallback =
            new ScanCallback() {
                @Override
                public void onScanResult(int callbackType, final ScanResult result) {
                    if (result.getDevice().getName() != null) {
                        Log.d("FOUND DEVICE", result.getDevice().getName());
                        String type = get_type(result.getDevice());
                        if (type != null) {
                            boolean found = false;
                            for (DeviceCell d : deviceCells) {
                                if (d.getDevice().getName().equals(result.getDevice().getName())) {
                                    found = true;
                                    break;
                                }
                            }

                            if (!found) {
                                Log.d("ADD DEVICE", result.getDevice().getName());
                                deviceCells.add(new DeviceCell(result.getDevice()));
                                myAdapter.notifyDataSetChanged();
                            }
                        }
                    }
                }
            };
    private static final long SCAN_PERIOD = 10000;

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private void scanLeDevice() {
        deviceCells.clear();
        myAdapter.notifyDataSetChanged();
        uiHandler.postDelayed(this::stopScan, SCAN_PERIOD);
        menuItem.setEnabled(false);
        mBluetoothLeScanner.startScan(mLeScanCallback);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public void stopScan() {
        menuItem.setEnabled(true);
        mBluetoothLeScanner.stopScan(mLeScanCallback);
    }

    @Override
    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public void onClickItem(DeviceCell deviceCell, BluetoothDevice device) {
        stopScan();

        DeviceInfo deviceInfo = new DeviceInfo(device, get_type(device));
        boolean found = false;
        for (DeviceInfo i: devs) {
            if (i.equals(deviceInfo)) {
                found = true;
                break;
            }
        }
        if (!found) {
            httpRequests.send_dev(deviceInfo);
        }

        finish();
    }
}

File: app/src/main/java/com/iomt/android/BLESearcher.kt
package com.iomt.android

import android.Manifest
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.GsonBuilder
import com.google.gson.reflect.TypeToken
import com.iomt.android.config.ConfigParser
import java.util.*

class BLESearcher : AppCompatActivity(), SavedDeviceAdapter.OnClickListener {
    private var uiHandler: Handler? = null
    private var menuItem: MenuItem? = null
    private val deviceCells: MutableList<DeviceCell> = ArrayList()
    private var myAdapter: SavedDeviceAdapter? = null
    private var mBluetoothLeScanner: BluetoothLeScanner? = null
    private var devTypes: List<DeviceType> = ArrayList()
    private var devs: List<DeviceInfo> = ArrayList()
    private lateinit var httpRequests: HTTPRequests
    private val configParser = ConfigParser()

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_searcher)
        val jwt = intent.getStringExtra("JWT")
        val userId = intent.getStringExtra("UserId")
        httpRequests = HTTPRequests(this, jwt, userId)
        val actionTypes = Action { args: Array<String?>? ->
            val builder = GsonBuilder()
            val gson = builder.create()
            val listOfDeviceTypes = object : TypeToken<ArrayList<DeviceType?>?>() {}.type
            devTypes = gson.fromJson(args!![0], listOfDeviceTypes)
        }
        httpRequests!!.getDeviceTypes(actionTypes)
        val actionDevs = Action { args: Array<String?>? ->
            val builder = GsonBuilder()
            val gson = builder.create()
            val listOfDevices = object : TypeToken<ArrayList<DeviceInfo?>?>() {}.type
            devs = gson.fromJson(args!![0], listOfDevices)
        }
        httpRequests!!.getDevices(actionDevs)
        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        mBluetoothLeScanner = bluetoothAdapter.bluetoothLeScanner
        uiHandler = Handler(Looper.getMainLooper())
        val toolbar = findViewById<View>(R.id.toolbar_searcher) as Toolbar
        setSupportActionBar(toolbar)
        val recyclerView = findViewById<View>(R.id.searcher_recycler_view) as RecyclerView
        recyclerView.setHasFixedSize(true)
        val layoutManager: RecyclerView.LayoutManager =
            LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        recyclerView.layoutManager = layoutManager
        myAdapter = SavedDeviceAdapter(LayoutInflater.from(this), deviceCells, this)
        recyclerView.adapter = myAdapter
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            requestPermissions(
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                PERMISSION_REQUEST_FINE_LOCATION
            )
        }
    }

    private fun getType(d: BluetoothDevice): String? {
        for (i in devTypes) {
            if (d.name.startsWith(i.prefix)) {
                return i.device_type
            }
        }
        return null
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == PERMISSION_REQUEST_FINE_LOCATION) {
            if (grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                menuItem!!.isEnabled = true
            } else {
                val builder = AlertDialog.Builder(this)
                builder.setTitle("Location not granted")
                builder.setMessage("Since location access has not been granted, this app will not be able to discover nearby bluetooth devices.")
                builder.setPositiveButton(android.R.string.ok, null)
                builder.show()
                menuItem!!.isEnabled = false
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        super.onCreateOptionsMenu(menu)
        val inflater = menuInflater
        inflater.inflate(R.menu.menu_searcher, menu)
        menuItem = menu.findItem(R.id.scan)
        return true
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.scan) {
            scanLeDevice()
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private val mLeScanCallback: ScanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            if (result.device.name != null) {
                Log.d("FOUND DEVICE", result.device.name)
                val type = getType(result.device)
                if (type != null) {
                    var found = false
                    for (d in deviceCells) {
                        if (d.device.name == result.device.name) {
                            found = true
                            break
                        }
                    }
                    if (!found) {
                        Log.d("ADD DEVICE", result.device.name)
                        deviceCells.add(DeviceCell(result.device))
                        myAdapter!!.notifyDataSetChanged()
                    }
                }
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    private fun scanLeDevice() {
        deviceCells.clear()
        myAdapter!!.notifyDataSetChanged()
        uiHandler!!.postDelayed({ stopScan() }, SCAN_PERIOD)
        menuItem!!.isEnabled = false
        mBluetoothLeScanner!!.startScan(mLeScanCallback)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    fun stopScan() {
        menuItem!!.isEnabled = true
        mBluetoothLeScanner!!.stopScan(mLeScanCallback)
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    override fun onClickItem(deviceCell: DeviceCell?, device: BluetoothDevice?) {
        stopScan()
        val deviceInfo = DeviceInfo(device!!, getType(device) ?: "UNKNOWN")
        var found = false
        for (i in devs) {
            if (i.equals(deviceInfo)) {
                found = true
                break
            }
        }
        if (!found) {
            httpRequests.sendDevice(deviceInfo)
            val stringDeviceConfig = httpRequests.getDeviceConfig(1)
            configParser.parseFromString(stringDeviceConfig ?: "")
            configParser.parse()
            Log.w(TAG, configParser.toString())
        }
        finish()
    }

    companion object {
        private const val TAG = "BLEScanner"
        private const val PERMISSION_REQUEST_FINE_LOCATION = 1
        private const val SCAN_PERIOD: Long = 10000
    }
}